### ISSUE TITLE
Regression on depends_on(between( in Lmod 8.6 ?

### ISSUE THREAD

**Issue by mboisson** [TACC/Lmod#564]
Hi Robert.
I think there has been a regression in the behavior of depends_on(between(...) in Lmod 8.6, which we are just testing now. 

With Lmod 8.4.16: 
```
[mboisson@narval3 ~]$ echo $LMOD_VERSION
8.4.16
[mboisson@narval3 ~]$ module show ipykernel/2022a | grep depends_on
depends_on(between("python","3.8<","<3.11"))
[mboisson@narval3 ~]$ module --terse list python
python/3.9.6
[mboisson@narval3 ~]$ module load ipykernel/2022a
[mboisson@narval3 ~]$ module --terse list ipykernel
ipykernel/2022a
``` 

With Lmod 8.6.11: 
```
[mboisson@build-node ~]$ echo $LMOD_VERSION
8.6.11
[mboisson@build-node ~]$ module show ipykernel/2022a | grep depends_on
depends_on(between("python","3.8<","<3.11"))
[mboisson@build-node ~]$ module --terse list python
python/3.9.6
[mboisson@build-node ~]$ module load ipykernel/2022a
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
The following dependent module(s) are not currently loaded: python/3.8.10 (required by: ipykernel/2022a)
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
[mboisson@build-node ~]$ module --terse list ipykernel
ipykernel/2022a
``` 

It still loads, but it shows a warning which should not be shown.

**Comment by rtmclay** [comment]
I am unable to reproduce your issue.  You can see what I did to run your test in `rt/depends_on/depends_on.tdesc` and the directory `rt/depends_on/mf/Core4`.
You can see my test on the latest on the master branch.

Please try repeating with a simple test similar to mine to see if you reproduce the same issue.

It might be that it depends having a cache file.  I'm not sure.

**Comment by mboisson** [comment]
Ok, I will do more tests with a simpler set of modules and see what triggers it. We also have modulerc files involved, that might be it. I'll try to pin it down more precisely and report back.

**Comment by mboisson** [comment]
Ok, here is a simpler setup: 
```
[mboisson@build-node ~]$ for f in $(find modulefiles -type f); do echo $f; echo ==================; cat $f; done
modulefiles/ipykernel/2022a.lua
==================
if convertToCanonical(LmodVersion()) >= convertToCanonical("8.2.9") then
        depends_on(between("python",'3.8<','<3.11'))
end
modulefiles/python/3.8.10.lua
==================
local root = "/cvmfs/soft.computecanada.ca/easybuild/software/2020/avx2/Core/python/3.8.10"

prepend_path("PATH", pathJoin(root, "bin"))
modulefiles/python/.modulerc.lua
==================
module_version("python/3.8.10", "3.8")
module_version("python/3.9.6", "3.9")
module_version("python/3.8", "default")
module_version("python/3.10.2", "3.10")
modulefiles/python/3.10.2.lua
==================
local root = "/cvmfs/soft.computecanada.ca/easybuild/software/2020/avx2/Core/python/3.10.2"

prepend_path("PATH", pathJoin(root, "bin"))
modulefiles/python/3.9.6.lua
==================
local root = "/cvmfs/soft.computecanada.ca/easybuild/software/2020/avx2/Core/python/3.9.6"

prepend_path("PATH", pathJoin(root, "bin"))
``` 

Interestingly, the error message changes when I move the `.modulerc.lua`. It then asks for the new default: 
```
[mboisson@build-node ~]$ echo $MODULEPATH
/home/mboisson/modulefiles
[mboisson@build-node ~]$ module load python/3.9
[mboisson@build-node ~]$ module load ipykernel
----------------------------------------------------------------------------------------------------------------------------The following dependent module(s) are not currently loaded: python/3.8.10 (required by: ipykernel/2022a)
----------------------------------------------------------------------------------------------------------------------------[mboisson@build-node ~]$ mv modulefiles/python/.modulerc.lua{,.bak}
[mboisson@build-node ~]$ module load ipykernel
----------------------------------------------------------------------------------------------------------------------------The following dependent module(s) are not currently loaded: python/3.10.2 (required by: ipykernel/2022a)
----------------------------------------------------------------------------------------------------------------------------
``` 

Here is our config: 
```
[mboisson@build-node ~]$ module --config |& grep default
Require Exact Match/no defaults    no
Allow extended default             yes
```

**Comment by mboisson** [comment]
By the way, on my system, your Core4 test also has the issue : 
```
[mboisson@build-node Lmod]$ module use /home/mboisson/git/Lmod/rt/depends_on/mf/Core4/
[mboisson@build-node Lmod]$ module avail

-------------------------------------------------------------------------------------------------------------------------------------------------------------------------- Your personal modules ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------
   ipykernel/2022a    python/3.8.10 (t)    python/3.9.6 (t,D)

  Où:
   t:  Tools for development / Outils de développement
   D:  Default Module

Utilisez "module spider" pour trouver tous les modules possibles.
Utilisez "module keyword key1 key2 ..." pour chercher tous les modules possibles qui correspondent à l'une des clés (key1, key2).


[mboisson@build-node Lmod]$ module load python/3.8
[mboisson@build-node Lmod]$ module load ipykernel
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
The following dependent module(s) are not currently loaded: python/3.9.6 (required by: ipykernel/2022a)
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

```

**Comment by rtmclay** [comment]
Thanks very much for the comment: 

_The issue happens when you have a version of python which is not the default version, and then you load ipykernel_

I am now able to reproduce the issue.  I do not have a fix yet.  I'll let you know when I do.

**Comment by rtmclay** [comment]
I have created the "testing" branch that fixes this issue for me.  Please test this branch to see if it works for you. Thanks very much for the bug report!

**Comment by mboisson** [comment]
Thanks. It looks like the testing branch fixes it.

It leaves a new warning behind that was not there before, but that one is admitedly caused by a bug in our modules, we have some `openmpi` module that have multiple conflicting dependencies :

```
depends_on("libfabric/1.11.0")
depends_on("libfabric/1.10.1")
```


that causes

```
The following dependent module(s) are not currently loaded: libfabric/1.11.0 (required by: openmpi/4.0.3)
```


but that's admitely correct to show, and we will need to figure out the issue in our module generation.

**Comment by rtmclay** [comment]
The testing branch has now been merged on the master branch and this fix is now in Lmod 8.6.15.
